### DISCUSSION TITLE
DAG inheritance and reusable tasks

### DISCUSSION THREAD

**Discussion by zeevo** [apache/airflow#16827] [General]
Hello Airflow!

What's the best way to achieve DAG inheritance or task composition when creating DAGs? For example, I want a common task to be available in the TaskFlow API on my DAG (defined with the `@dag` decorator) always. I have many DAGs and some common tasks between them such as generic Error handling, and instead of copy and pasting my `@task` declaration around different TaskFlows, I would love to be able to extend a common DAG so that the task is always available.

How does one achieve this in Airflow?

**Comment by zeevo** [comment]
Any thought or comments on bringing these concepts into Airflow? Maybe extendable DAGs in some way?

**Comment by potiuk** [comment]
Why not declaring your common functions in shared python package and importing them as needed ? Is there any problem with that? DAGs are python code and you can use the usual DRY mechanisms to extract common code, as you'd use in Python.

Same with DAG - you can define a "base DAG" with common tasks added, then perform a factory function that will return you the base dag with all the common tasks. Another approach with might work is deepcopy a base DAG every time you want to use it as a base and use that deepcopy as a new DAG to start with and add new tasks.

Is this something you can't do for any reason?

**Reply by josh-fell** [reply]
@zeevo Perhaps you may find this talk incorporating DRY principles into Airflow development from the Airflow Summit this year to be useful: https://www.youtube.com/watch?v=ewPj1L4I68c
